lib/pcr.c

```diff
@@ -237,26 +237,114 @@ bool pcr_print_pcr_struct_le(TPML_PCR_SELECTION *pcr_select, tpm2_pcrs *pcrs) {
     return result;
 }
 
-bool pcr_print_pcr_struct(TPML_PCR_SELECTION *pcr_select, tpm2_pcrs *pcrs) {
+bool pcr_fwrite_serialized(const TPML_PCR_SELECTION *ppcr_select,
+                           const tpm2_pcrs *ppcrs, FILE *output_file) {
+
+    TPML_PCR_SELECTION pcr_selections = *ppcr_select;
+    // Make sure the file content is written in little endian format
+    pcr_selections.count = htole32(pcr_selections.count);
+    UINT32 i;
+    for (i = 0; i < le32toh(pcr_selections.count); i++) {
+        pcr_selections.pcrSelections[i].hash = htole16(
+                pcr_selections.pcrSelections[i].hash);
+    }
```

Comment:
Rather than doing yet another binary format, can't we just make this YAML?

Plus, since in general things are host endianess or big endianness in this code base, BE over LE.

Reply:
This particular binary format ("serialized") is currently the format used by `tpm2_quote` to output the selected list of PCR's.  In contrast, the binary format used by `tpm2_pcrread` is currently equivalent to the proposed "values" format.  Since the existing output format behavior of the two commands are not are different and not selectable, it is not possible to compare the output files (of the same PCR values, for example) directly to see if they are the same or not.

For backward compatibility, this patch doesn't change the default behavior of the two commands, while allowing the users to choose the specific output formats to use.  This helps alleviate the comparison problems described above.

In other words, this patch hasn't created "yet another binary format", but rather makes the existing binary formats available for selection and reuse. To that end, I have not changed the endianess of the existing binary formats.

We can certainly support additional formats, such as YAML, but I think that belongs to a future enhancement.

Reply:
Ahh that knocks the dust off my brain. I remember @idesai and I having a discussion on this way back when we noticed that the tools had differing PCR formats. So this supports an interchange between those tools. +2 from me.